### ISSUE TITLE
update eslint:recommended and eslint-config-eslint.

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#14673]
**The version of ESLint you are using.**
latest

**The problem you want to solve.**
update eslint:recommended and eslint-config-eslint.

**Your take on the correct solution to problem.**
* y: enable
* n: not enable
* ?: not sure

Name | Added in eslint version | eslint:recommended | eslint-config-eslint | Description
----- | ----- | ----- | ----- | -----
[default-case-last](https://eslint.org/docs/rules/default-case-last) | 7.0.0-alpha.0 | ? | y | enforce default clauses in switch statements to be last
[default-param-last](https://eslint.org/docs/rules/default-param-last) | 6.4.0 | ? | y | enforce default parameters to be last
[function-call-argument-newline](https://eslint.org/docs/rules/function-call-argument-newline) | 6.2.0 | n | y('consistent') | enforce line breaks between arguments of a function call
[grouped-accessor-pairs](https://eslint.org/docs/rules/grouped-accessor-pairs) | 6.7.0 | n | y | require grouped accessor pairs in object literals and classes
[id-denylist](https://eslint.org/docs/rules/id-denylist) | 7.4.0 | n | n | disallow specified identifiers
[no-constructor-return](https://eslint.org/docs/rules/no-constructor-return) | 6.7.0 | y | y | disallow returning value from constructor
[*no-dupe-else-if](https://eslint.org/docs/rules/no-dupe-else-if) | 6.7.0 | y | y | disallow duplicate conditions in if-else-if chains
[*no-import-assign](https://eslint.org/docs/rules/no-import-assign) | 6.4.0 | y | y | disallow assigning to imported bindings
[no-loss-of-precision](https://eslint.org/docs/rules/no-loss-of-precision) | 7.1.0 | y | y | disallow literal numbers that lose precision
[no-nonoctal-decimal-escape](https://eslint.org/docs/rules/no-nonoctal-decimal-escape) | 7.14.0 | y | y | disallow `\8` and `\9` escape sequences in string literals
[no-promise-executor-return](https://eslint.org/docs/rules/no-promise-executor-return) | 7.3.0 | y | y | disallow returning values from Promise executor functions
[no-restricted-exports](https://eslint.org/docs/rules/no-restricted-exports) | 7.0.0-alpha.0 | n | n | disallow specified names in exports
[*no-setter-return](https://eslint.org/docs/rules/no-setter-return) | 6.7.0 | y | y | disallow returning values from setters
[no-unreachable-loop](https://eslint.org/docs/rules/no-unreachable-loop) | 7.3.0 | y | y | disallow loops with a body that allows only one iteration
[no-unsafe-optional-chaining](https://eslint.org/docs/rules/no-unsafe-optional-chaining) | 7.15.0 | y | y | disallow use of optional chaining in contexts where the `undefined` value is not allowed
[no-useless-backreference](https://eslint.org/docs/rules/no-useless-backreference) | 7.0.0-alpha.0 | y | y | disallow useless backreferences in regular expressions
[prefer-exponentiation-operator](https://eslint.org/docs/rules/prefer-exponentiation-operator) | 6.7.0 | n | y | disallow the use of `Math.pow` in favor of the `**` operator
[prefer-regex-literals](https://eslint.org/docs/rules/prefer-regex-literals) | 6.4.0 | n | y | disallow use of the `RegExp` constructor in favor of regular expression literals

is it too late to consider adding this to eslint v8.0.0? :-(

**Comment by nzakas** [comment]
It’s not clear to me what exactly you are suggesting should change. Are you saying that rules with “y” in the eslint:recommended table should be *added*? I’m just not sure how to read this table. 

Keep in mind that recommended rules must flag potential errors or suggest better ways of doing things. We can’t recommend any rules that are stylistic.

**Comment by aladdin-add** [comment]
> Are you saying that rules with “y” in the eslint:recommended table should be added?

yes! updated the issue to make it clearer.

**Comment by nzakas** [comment]
Thanks, that helps. I think a lot of these rules fall under the category of stylistic, so I’m not in favor of adding them to eslint:recommended. 

The ones I think make sense to add are:

* no-import-assign
* no-loss-of-precision
* no-nonoctal-decimal-escape
* no-promise-executor-return
* no-setter-return
* no-unreachable-loop
* no-unsafe-optional-chaining
* no-useless-backreference

I don’t care too much about eslint-config-eslint. Historically, we have tried to enable as many rules as possible so we are constantly testing them. As such, I’m fine with adding whatever people want.

**Comment by mdjermanovic** [comment]
I vote for these four:

* no-loss-of-precision
* no-unsafe-optional-chaining
* no-useless-backreference
* no-nonoctal-decimal-escape

The first three flag errors in the code, and I can't think of any reason to not enable each.

no-nonoctal-decimal-escape flags the use of a legacy syntax that can be easily avoided.

---------

no-unreachable-loop

This can be problematic because by default it flags a pattern that checks if an object has enumerable properties:

```js
for (const key in obj) {
  has = true;
  break;
}
```

And a pattern to check if there's anything in a sequence:

```js
for (const value of iterable) {
  has = true;
  break;
}
```

---------

no-promise-executor-return

There are some patterns that would be reported by this rule:

* https://github.com/eslint/eslint/issues/12640#issuecomment-561900061
* https://github.com/eslint/eslint/issues/13668

---------

Rest of the rules marked as "?" or "y":

* default-case-last: while very uncommon, the code it flags is valid.
* default-param-last: this is only a best practice, the code it flags is valid.
* no-constructor-return: I believe most codebases should enable this rule to catch errors in the code, but there are valid patterns where constructor returns a value.
* no-dupe-else-if: this rule is already in eslint:recommended
* no-import-assign: this rule is already in eslint:recommended
* no-setter-return: this rule is already in eslint:recommended

**Comment by nzakas** [comment]
Happy to go along with @mdjermanovic’s suggestions.

**Comment by aladdin-add** [comment]
sounds good!

I can make 2 PRs: one for the `eslint:recommended`; one for `eslint-config-eslint`.

**Comment by snitin315** [comment]
Sounds good to me 👍🏻

**Comment by btmills** [comment]
I also agree with @mdjermanovic for `eslint:recommended`. For `eslint-config-eslint`, similar to @nzakas I’m fine trying to enable the ones you suggested, and we can disable one if it’s too difficult to adopt.
